[WIP] convert domain_memberships to nested field

sravfeyn opened the pull request:
## Product Description
User Elasticsearch mapping changes https://dimagi-dev.atlassian.net/browse/USH-3611

## Technical Summary

Makes Elasticsearch `domain_membership(s)` fields on user index a nested field. Updates relevant queries to use nested filters

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
NA

## Safety Assurance
Tested locally

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage
Tests added
<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->



### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations
This includes an index change, so to roll back the previous index needs to exist or be restored.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

sravfeyn committed:
Add test for user_ids_at_locations

sravfeyn committed:
remove unused methods/tests

sravfeyn committed:
Refactor to add test

sravfeyn committed:
add user filters back

sravfeyn committed:
fix test; lint

sravfeyn committed:
convert domain_memberships to nested field

sravfeyn commented:
@esoergel I have opened this as draft just for an initial validation of the approach. (no full review required as I plan to update the PR)

sravfeyn commented:
@esoergel I have opened this as draft just for an initial validation of the approach. (no full review required as I plan to update the PR)

esoergel commented on corehq/apps/users/migrations/0004_rm_role_id_from_admins.py:
```diff
@@ -14,13 +14,15 @@ def fix_users(apps, schema_editor):
 
 
 def _get_admins_with_roles():
```
You can just delete the whole `RunPython` operation for this migration - it's not relevant anymore

sravfeyn replied:
Sure! This would be a no-op migration, then, yeah? I guess I will keep the code around just in case?

esoergel replied:
To avoid confusing the migrations framework, the migration file still needs to be there, but you can just set `operations = []` like in this file:

https://github.com/dimagi/commcare-hq/blob/d9e67f9fde452ac5b3682af839f5b4583404436e/corehq/apps/users/migrations/0010_populate_usersinvitation.py

sravfeyn commented on corehq/apps/es/queries.py:
```diff
@@ -75,13 +84,31 @@ def search_string_query(search_string, default_fields):
+
+    if nested_fields:
+        return BOOL_CLAUSE(
+            SHOULD_CLAUSE([default_field_query] + nested_queries)
```
Yes, it does behave the same. I have added a [test](https://github.com/dimagi/commcare-hq/pull/33694/files#diff-ec315c6fbc57958b15ac0681b8f7cd4cd696a18ba451e3387522365e8aa6e57eR164) to test this.

Regarding storing domains for webusers, of course we can do it, but I thought we would try to maintain partiy between ES and couch as much. This info is already captured int the `domain_memberships`, so why duplicate the domain list to another property since we are still able to query well.

esoergel replied:
That test still uses single term queries though - what I mean is that you can currently go to the report and type `testuser34 esoergel` and see a result where the first term matches a username and the second term matches a domain.  What this query would do in that case (I believe) is look for a match for `testuser34 esoergel` across username, first and last name, and phone number, and find no matches, then look for a match in `domain_memberships.domain` and find no matches.

That util is only used to support the admin report, and I don't think we need to worry too much about changing behavior there.  It might also be fine to just drop `domain` from the list of fields checked against, since if you already know the domain, you'd probably just look at that domain's users list.  Though if you're doing a partial match on the domain maybe it would be nice.

sravfeyn replied:
Ah, got it, I will look into this.

sravfeyn commented on corehq/apps/es/users.py:
```diff
@@ -268,3 +271,11 @@ def missing_or_empty_metadata_property(property_name):
     )
+
+
+def domain_memberships_filter(field_name, value):
```
That's true. Though, the intention of the method was only to provide a shortcut to the long nested filter to be DRY. I didn't intend it to include a domain filter. We could of course make another method for that, but it wasn't necessary for the code I that I had to modify.

sravfeyn commented on corehq/apps/es/mappings/user_mapping.py:
```diff
@@ -117,7 +117,7 @@
         "domain_membership": {
             "dynamic": False,
-            "type": "object",
+            "type": "nested",
```
Yes, that is the next thing I was going to do!

sravfeyn replied:
[9e7b4095059703af19b35b1ae0711eb256d1dea3](https://github.com/dimagi/commcare-hq/pull/33726/commits/9e7b4095059703af19b35b1ae0711eb256d1dea3)

sravfeyn replied:
I am wondering whether this change is actually worth it given the confusion it creates over how the couch and ES docs differ for mobile users doesn't seem justified for the little code cleanup it offers. What do you think?

esoergel replied:
That's a fair consideration.  I'm hoping to bring things closer together to avoid accidentally introducing code that works for only one type of user.  For what it's worth, the `CommCareUser` class [already has `domains` and `domain_memberships` properties of this nature](https://github.com/dimagi/commcare-hq/blob/d9e67f9fde452ac5b3682af839f5b4583404436e/corehq/apps/users/models.py#L673-L679) for presumably the same reason.

The cleanup I'm referring to is stuff like this:
https://github.com/dimagi/commcare-hq/blob/d9e67f9fde452ac5b3682af839f5b4583404436e/corehq/apps/es/users.py#L199-L207

Sure it's not hard to work around in code, just seems annoying and kind of silly.  Though I do agree that it's nice to have the couch and ES docs similar in general.  Another possibility is to do both.  That is, we could send mobile workers with both `domain_membership` and `domain_memberships`, but only have the latter indexed in the mapping.  Then we can have data indexed in a way that's straightforward to query, but still have the expected fields populated in the return value.

I don't know, maybe we should invite comment from the broader team?

sravfeyn committed:
Index domain_membership as domain_memberships in ES for mobile users

sravfeyn committed:
send domain_memberships without deleting domain_membership

sravfeyn committed:
Make user_data_es doubly nested

sravfeyn commented:
@esoergel I have added user_data changes as well now. https://github.com/dimagi/commcare-hq/pull/33726/commits/7555c3dbdbfc639432609404958a1c70a4ef7c27 (Also rebased this off of your `es/sql-user-data` branch)

sravfeyn commented on corehq/apps/es/tests/test_user_es.py:
```diff
@@ -59,3 +62,12 @@ def test_chained_user_data_queries_with_only_one_match(self):
                              .values_list('username', flat=True))
         self.assertEqual(direwolf_families, [])
+
+    def test_missing_key(self):
```
Todo: I will add webuser tests as well.

esoergel commented on corehq/apps/es/users.py:
```diff
@@ -107,12 +108,24 @@ def _from_dict(self, user_dict):
         user_dict['user_data_es'] = []
         if user_dict.get('base_doc') == 'CouchUser' and user_dict['doc_type'] == 'CommCareUser':
```
Exciting!  I think you can drop the second half of this conditional now - that'll be what's needed to get web users populated properly

sravfeyn replied:
I have noticed that the webuser user_data updates are still not working [per this test you have added](https://github.com/dimagi/commcare-hq/pull/33688/files#diff-7cfcbcfb34b82f54d196ac035c3839a8b61a164ea16efe51fc59fcea7e1369daR73). Are you planning to fix this in your PRs? I suppose that would be required to add Webuser tests in this PR (with multiple domain memberships)

esoergel commented on corehq/apps/es/users.py:
```diff
@@ -261,11 +267,14 @@ def _missing_user_data_value(property_name):
     return queries.nested(
-        'user_data_es',
+        "user_data_es",
         filters.AND(
```
nit: this `AND` is no longer necessary, as there's a single clause inside it.  Though I think `_missing_user_data_value` and `_missing_user_data_property` would probably be simpler if combined, especially once a `domain` filter is applied.

sravfeyn replied:
Removed `AND`

esoergel reviewed (commented):
looking good!

sravfeyn commented on corehq/apps/es/users.py:
```diff
@@ -236,10 +234,15 @@ def is_active(active=True):
     return queries.nested(
-        'user_data_es',
+        "user_data_es",
         filters.AND(
```
Yup! I will add a domain kwarg.

sravfeyn committed:
add domain kwarg

sravfeyn committed:
make migration no-op

sravfeyn committed:
add field back

sravfeyn commented:
@esoergel 

I have noticed that the webuser user_data updates are not yet working [per this test you have added](https://github.com/dimagi/commcare-hq/pull/33688/files#diff-7cfcbcfb34b82f54d196ac035c3839a8b61a164ea16efe51fc59fcea7e1369daR73). Are you planning to fix this in your PRs? I suppose that would be required to add Webuser tests in this PR (with multiple domain memberships)

esoergel commented:
@sravfeyn 

> I have noticed that the webuser user_data updates are not yet working [per this test you have added](https://github.com/dimagi/commcare-hq/pull/33688/files#diff-7cfcbcfb34b82f54d196ac035c3839a8b61a164ea16efe51fc59fcea7e1369daR73). Are you planning to fix this in your PRs? I suppose that would be required to add Webuser tests in this PR (with multiple domain memberships)

Good point - post https://github.com/dimagi/commcare-hq/pull/33688/ web users are a little weird - it should work fine once initialized, so once the management command has been run (or if user data has been lazily initialized for applicable users and domains), things should be ready.  I added [a test demonstrating that](https://github.com/dimagi/commcare-hq/pull/33688/commits/de687004867a5a37580b74d8b6de6e1ccf430c68).  I just filled out the description of that PR with an explainer of the rollout I had in mind.

For the purposes of this PR, you should be able to mimic the final behavior as I did in that test by initializing user data for each domain before introducing fields that need to be different.  The behavior should make a bit more sense once I remove the lazy initialization code.

esoergel commented:
Regarding reindexing options - Option 1 sounds reasonable to me.  I was hoping we'd be able to just copy unknown and admin users over unchanged, but you're right that we'd need to fiddle with the domain memberships thing, looks like this is what gets sent to ES now:
https://github.com/dimagi/commcare-hq/blob/d9e67f9fde452ac5b3682af839f5b4583404436e/corehq/pillows/user.py#L34-L42

Can't wait till we figure out how to remove unknown users - I'm not even sure we use them for anything useful.

sravfeyn committed:
Merge es/sql-user-data

sravfeyn committed:
pass web_user data

sravfeyn committed:
rewrite login_as_user query

esoergel commented on corehq/apps/es/users.py:
```diff
@@ -233,7 +233,13 @@ def is_active(active=True):
+        return _user_data_filter(key, value, domain=domain)
+
 
+def _user_data_filter(key, value, domain=None):
```
This uses `queries.match`, which seemed like an odd choice, assuming we're generally looking for exact matches, not similar terms.  Looks like I made that decision deliberately back in
https://github.com/dimagi/commcare-hq/pull/28667
Though I'm not sure I'm swayed by my reasoning at the time.  What do you think?  Out of scope for this PR sure, but I'm wondering whether switching from a `term` to a `match` results in functional changes for the login-as queries here.

sravfeyn replied:
Yeah, I had similar thoughts, but kept it as it is to avoid any functional regressions. It could be a separate task for latter.

esoergel commented on corehq/apps/cloudcare/esaccessors.py:
```diff
@@ -39,28 +39,13 @@ def login_as_user_query(
-            )
-        )
+            usernames.append('default')
+        user_es = user_es.user_data('login_as_user', usernames)
```
Should this be domain-scoped?  Like, could I set `login_as_user` on another domain to the username of someone I shouldn't be able to access on the main domain, and then access them anyways?

sravfeyn replied:
That's a good point, I guess that behaviour is indeed possible now, but we can fix that now (that we have nested fields for domain_memberships)?

esoergel replied:
Ah right yeah that makes sense

dannyroberts commented:
Converted to draft since this is still a WIP to prevent accidental merge

sravfeyn committed:
Add ES migration to create new users index

sravfeyn committed:
Add ability to specify script to ES reindex command

sravfeyn commented on corehq/apps/es/client.py:
```diff
@@ -445,6 +450,14 @@ def reindex(
+                reindex_body["script"] = json.loads(f.read())
+        if source_query_json:
+            with open(source_query_json) as f:
+                reindex_body["source"]["query"] = json.loads(f.read())
```
We will be able to run below command to reindex unknownusers from old index to new index.

```
/manage.py elastic_sync_multiplexed start users --script_json=corehq/apps/es/management/commands/scripts/add_domain_memberships.json --source_query_json=corehq/apps/es/management/commands/scripts/query_unknown_users.json
```

sravfeyn committed:
Revert "Add ability to specify script to ES reindex command"

This reverts commit 3afbc0cc6d9129941bb14ac29eb93f3f441422c7.

sravfeyn committed:
Copy unknown users

sravfeyn commented:
Regarding reindex process, I have chatted with Amit (to align with SaaS). We agreed that we will be following the old reindex process to roll this out. (The timelines and logistics are discussed on the [ticket](https://dimagi-dev.atlassian.net/browse/USH-3611), we can do the technical discussion here)

- On Production: Do a pre deploy index rollout
  - Setup a prerelease branch
  - Create the index and populate using ptop_reindexer_v2
  - Run `copy_es_unknownusers` management command to copy unknown/admin users
- For other envs: Rollout as part of deploy
  - Merge and deploy this branch
  - The create-index migration automatically creates the new index
  - Another migration (to be added) will run the newly created `copy_es_unknownusers` management command
- For 3rd party envs with large indices
  - #1 will be published as a changelog prior to #2

sravfeyn committed:
fix data accssor
